# Input
* Maskfile: NP.chr4.fa.gz
* Diploperennis/luxurians ancestral reference genome: diplux.2.fasta.gz
* Recombination rate map: ogut.chr4.txt

# Scripts
* Relate script: run_relate.sh
* Convert Relate output to .trees: convertTree.sh
* Calculate SFS/AFS on Relate output and on observed VCF: calcSFS.py
